includes/ProductSets/Sync.php

```diff
@@ -0,0 +1,382 @@
+	 *
+	 * @param int $term_id Product Cat Term ID.
+	 */
+	public function maybe_sync_product_set_on_product_cat_remove( $term_id ) {
+
+		$product_sets = $this->get_product_cat_sets( $term_id );
+		if ( empty( $product_sets ) ) {
+			return;
+		}
+
+		foreach ( $product_sets as $product_set_id ) {
+			$this->maybe_sync_product_set( $product_set_id );
+		}
+	}
+
+
+	/**
+	 * Call API to remove Product Set from FB
+	 *
+	 * @since 2.1.5
+	 *
+	 * @param int $product_set_term_id Product Set Term ID.
+	 */
+	public function sync_remove_product_set( $product_set_term_id, $taxonomy ) {
+
+		if ( 'fb_product_set' !== $taxonomy ) {
+			return;
+		}
+
+		// get product set FB ID
+		$fb_product_set_id = get_term_meta( $product_set_term_id, \WC_Facebookcommerce_Integration::FB_PRODUCT_SET_ID, true );
+		if ( empty( $fb_product_set_id ) ) {
+			return;
+		}
+
+		do_action( 'fb_wc_product_set_delete', $fb_product_set_id );
+	}
+
+
+	/** Utility methods *******************************************************************************************/
+
+
+	/**
+	 * Maybe sync product set
+	 *
+	 * @since 2.1.5
+	 *
+	 * @param int $product_set_id FB Product Set Term ID.
+	 */
+	public function maybe_sync_product_set( $product_set_id ) {
+
+		// get Product Set
+		$term = get_term( $product_set_id );
+		if ( empty( $term ) ) {
+			return;
+		}
+
+		// get categories
+		$product_cat_ids = get_term_meta( $product_set_id, $this->categories_field, true );
+
+		// get products for the taxonomies
+		$products      = array();
+		$products_args = array(
+			'fields'         => 'ids',
+			'post_type'      => 'product',
+			'posts_per_page' => -1,
+			'tax_query'      => array(
+				'relation' => 'OR',
+				array(
+					'taxonomy' => 'fb_product_set',
+					'field'    => 'term_taxonomy_id',
+					'terms'    => array( $product_set_id ),
+					'operator' => 'IN',
+				),
+				array(
+					'taxonomy' => 'product_cat',
+					'field'    => 'term_taxonomy_id',
+					'terms'    => $product_cat_ids,
+					'operator' => 'IN',
+				),
+			),
+		);
+		$product_ids   = get_posts( $products_args );
+
+		// formats filter value
+		foreach ( $product_ids as $product_id ) {
+			array_push(
+				$products,
+				array( 'retailer_id' => array( 'eq' => self::PRODUCT_INDEX_PREFIX . $product_id ) )
+			);
+		}
+		$data = array(
+			'name'   => $term->name,
+			'filter' => wp_json_encode( array( 'or' => $products ) ),
+		);
+
+		do_action( 'fb_wc_product_set_sync', $data, $product_set_id );
+	}
+
+
+	/**
+	 * Return the list of product sets of a given product category term
+	 *
+	 * @since 2.1.5
+	 *
+	 * @param string $product_cat_id product cat Term ID.
+	 *
+	 * @return array
+	 */
+	protected function get_product_cat_sets( $product_cat_id ) {
+
+		$args = array(
+			'fields'     => 'ids',
+			'taxonomy'   => 'fb_product_set',
+			'hide_empty' => false,
+			'meta_query' => array(
+				array(
+					'key'     => 'wc_product_cats',
+					'value'   => sprintf( ':%d;', $product_cat_id ),
```

Comment:
Potentially heavy query here, does this run frequently?

Reply:
Seems like it doesn't, but aren't we missing a % for the LIKE comparison?

Reply:
It runs only on _delete_product_cat_ and Product _save_post_ actions. 
The % is added automatically when using LIKE: 
https://core.trac.wordpress.org/browser/tags/3.8.1/src/wp-includes/meta.php#L842